api/api.go

```diff
@@ -0,0 +1,340 @@
+	d := json.NewDecoder(postResp.Body)
+	if err := d.Decode(&resp); err != nil {
+		return fmt.Errorf("decode: %v", err)
+	}
+	if !resp.Success {
+		return fmt.Errorf("registration failed: %v", resp.message())
+	}
+	api.token = resp.Token
+	return nil
+}
+
+func (api *CacophonyAPI) getFileFromJWT(jwt, path string) error {
+	// Create the file
+
+	out, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Get the data
+	resp, err := http.Get(api.serverURL + "/api/v1/signedUrl?jwt=" + jwt)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Check server response
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status: %s", resp.Status)
+	}
+
+	// Writer the body to file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetFilesFromSchedule will get all files from the IDs in the schedule and save to disk.
+func (api *CacophonyAPI) GetFilesFromSchedule(schedule Schedule, fileFolder string) error {
+	err := os.MkdirAll(fileFolder, 0755)
+	if err != nil {
+		return err
+	}
+
+	for _, fileID := range schedule.AllSounds {
+		err := api.GetFile(fileID, filepath.Join(fileFolder, strconv.Itoa(fileID)))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetFile will download a file from the files api and save to disk
+func (api *CacophonyAPI) GetFile(fileID int, path string) error {
+	if _, err := os.Stat(path); err == nil {
+		return nil
+	}
+	buf := new(bytes.Buffer)
+
+	req, err := http.NewRequest("GET", api.serverURL+"/api/v1/files/"+strconv.Itoa(fileID), buf)
+	req.Header.Set("Authorization", api.token)
+	client := new(http.Client)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	var fr fileResponse
+	d := json.NewDecoder(resp.Body)
+	if err := d.Decode(&fr); err != nil {
+		return err
+	}
+	return api.getFileFromJWT(fr.Jwt, path)
+}
+
+type fileResponse struct {
+	Jwt string "jwt"
+}
+
+// GetSchedule will get the audio schedule
+func (api *CacophonyAPI) GetSchedule() (Schedule, error) {
+	log.Println("Getting new schedule")
+	req, err := http.NewRequest("GET", api.serverURL+"/api/v1/schedules", nil)
+	req.Header.Set("Authorization", api.token)
+	client := new(http.Client)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return Schedule{}, err
+	}
+	defer resp.Body.Close()
+	/*
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		bodyString := string(bodyBytes)
+		//log.Printf("status code: %d, body:\n%s", resp.StatusCode, )
+		//return "", errors.New("non 200 status code")
+	*/
+	var sr scheduleResponse
+	d := json.NewDecoder(resp.Body)
+	if err := d.Decode(&sr); err != nil {
+		return Schedule{}, err
+	}
+
+	return sr.Schedule, nil
+}
+
+type scheduleResponse struct {
+	//Success  bool     "succes"
+	Schedule Schedule "schedule"
+}
+
+// Schedule fo the audio lures
+type Schedule struct {
+	Combos      []Combo "combos"
```

Comment:
Are all these struct tags actually working? I'm pretty sure they're supposed to take this form: `json:"combos"` in order to work for JSON marshalling..